### ISSUE TITLE
cannot batch correct between technologies, e.g. smartseq and 10X

### ISSUE THREAD

**Issue by aopisco** [YosefLab/scVI#627]
I'm using the following code, which a cp from the tutorial:

```
import scvi
from scvi.dataset import AnnDatasetFromAnnData
from scvi.inference import UnsupervisedTrainer
from scvi.models.vae import VAE
from typing import Tuple


def compute_scvi_latent(
    adata: sc.AnnData,
    scVI_model_name: str,
    n_latent: int = 10,
    n_epochs: int = 150,
    lr: float = 1e-5,
    use_batches: bool = True,
    use_cuda: bool = True,
) -> Tuple[scvi.inference.Posterior, np.ndarray]:
    
    """
    Train and return a scVI model and sample a latent space
    
    :param adata: sc.AnnData object non-normalized
    :param n_latent: dimension of the latent space
    :param n_epochs: number of training epochs
    :param lr: learning rate
    :param use_batches
    :param use_cuda
    :return: (scvi.Posterior, latent_space)
    
    """
    print(scVI_model_name)
    print('Convert easily to scvi dataset')
    scviDataset = AnnDatasetFromAnnData(adata,batch_label = 'batch')
    print(np.unique(scviDataset.batch_indices))
    print(scviDataset.n_batches)

    print('Training a model...')
    print('VAE')
    vae = VAE(
        scviDataset.nb_genes,
        n_batch=scviDataset.n_batches * use_batches,
        n_hidden=128, n_latent=30, n_layers=5, dispersion='gene'
    )
    
    print('UnsupervisedTrainer')
    trainer = UnsupervisedTrainer(vae, scviDataset, train_size=.99, use_cuda=use_cuda,
                                  n_epochs_kl_warmup=None,n_iter_kl_warmup=int(128*5000/400))
    
    trainer.train(n_epochs=n_epochs, lr=lr)
    ####
    
    torch.save(trainer.model.state_dict(), scVI_model_name)

    print('Extracting latent space')
    posterior = trainer.create_posterior(
        trainer.model, scviDataset, indices=np.arange(len(scviDataset))
    ).sequential()

    latent, _, _ = posterior.get_latent()

    return posterior, latent
```
I defined the batch as:

```
adata.obs['batch'] = adata.obs['method'].astype(str) + '_' +adata.obs['tissue'].astype(str)

batch_dict_keys = list(set(adata.obs['batch']))
batch_dict_values = range(len(batch_dict_keys))

batch_dict = {batch_dict_keys[i]: batch_dict_values[i] for i in range(len(batch_dict_keys))}

adata.obs['batch_categories'] = adata.obs['batch'].copy()
adata.obs = adata.obs.replace({'batch': batch_dict})
```
where method is either smartseq2 or 10X. Then

```
n_epochs_all = None
n_epochs = 150 if n_epochs_all is None else n_epochs_all

scvi_posterior, scvi_latent = compute_scvi_latent(
    adata, n_epochs=n_epochs, n_latent=30,
    use_batches = True, use_cuda=True, scVI_model_name = 'lemur-scVI-'+celltypesubset+'.pkl'
)
```
 followed by

```
sc.pp.neighbors(adata, n_neighbors=20, n_pcs=30, use_rep="X_scvi")
sc.tl.umap(adata)
sc.tl.leiden(adata, resolution=5, key_added="leiden_scvi")

sc.pl.umap(adata,color=['method','tissue','individual'],ncols=1)
```

which yields
<img width="621" alt="Screen Shot 2020-03-23 at 7 28 33 AM" src="https://user-images.githubusercontent.com/25906087/77327066-eee56180-6cd7-11ea-9ac2-2a5138170811.png">

any suggestions?

**Comment by adamgayoso** [comment]
What were the results of
``` python
print(np.unique(scviDataset.batch_indices))
print(scviDataset.n_batches)
```
?

**Comment by adamgayoso** [comment]
It's unclear exactly from just reading the code, but I think the problem may be that `adata.obs['batch']` contains strings and not ints.

**Comment by aopisco** [comment]
<img width="918" alt="Screen Shot 2020-03-23 at 8 30 50 AM" src="https://user-images.githubusercontent.com/25906087/77333422-a0889080-6ce0-11ea-935e-cf62034b4d78.png">

**Comment by adamgayoso** [comment]
Ok I would still inspect manually that `scviDataset.batch_indices` contains what you want (and as ints).

Other reasons this can occur:

1. The `lr` is `1e-5`, this seems really low, so the model probably is underfitting the data. I'd use `1e-3` at a minimum.
2. How many genes are you using? Unless you need all the genes for some reason, performance will be better if you cut down to a few thousand genes.
3. 5 hidden layers means there are even more parameters to train, I'd run for 400 epochs at a minimum, unless you're manually inspecting the loss. I'm not even sure 5 hidden layers is necessary.
4. 30 dimensions for the latent space also seems a bit high. scVI should more efficiently compress the data than a PCA, and it's possible with more dimensions here that disentangling the batch is harder.

**Comment by adamgayoso** [comment]
@aopisco I'm going to close this issue, but feel free to reopen if you have any other questions or comments!

**Comment by romain-lopez** [comment]
@aopisco After all these things, another point is that I think we performed some type of gene length correction in our analysis for merging some SS2 and 10x data? I think this is detailed in our scANVI manuscript, I'm sure @chenlingantelope can confirm / help / explain on that end.
